### PR TITLE
os::cmd improvements

### PR DESCRIPTION
**Pull request by stevekuznetsov**
#### Better naming conventions in `os::cmd` output
Tests will be internally named better, with the test declaration line looking like:
```
test/cmd/edit.sh:18: executing 'OC_EDITOR=cat oc edit pod/hello-openshift' expecting success and text 'Edit cancelled'...
```
with options:
```
path/to/file.sh:line: executing <cmd> expecting (success|failure|exit code <code>)(and (not) text <text>)?
```

Output will look like:
```
SUCCESS: test/cmd/edit.sh:18: executing 'OC_EDITOR=cat oc edit pod/hello-openshift' expecting success and text 'Edit cancelled'
FAILURE: test/cmd/edit.sh:18: executing 'OC_EDITOR=cat oc edit pod/hello-openshift' expecting success and text 'Edit cancelled': the command returned the wrong error code; the output content test failed
```
with similar options

#### `os::cmd` reports time elapsed for tests, with sub-second precision so we can see something meaningful for short tests
```
SUCCESS after 0.038322103s: test/cmd/edit.sh:16: executing 'OC_EDITOR=cat oc edit pod/hello-openshift' expecting success and text 'Edit cancelled'
```

#### `os::cmd` only uses `local` variables now to make it impossible for us to interfere with other variables set in the process

@deads2k @liggitt  PTAL

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
[test]

**Comment by stevekuznetsov** [comment]
[test]

**Comment by stevekuznetsov** [comment]
flake [here](Get https://copr.fedoraproject.org/coprs/maxamillion/origin-next/repo/epel-7/maxamillion-origin-next-epel-7.repo: dial tcp 209.132.184.54:443: no route to host):
```
Get https://copr.fedoraproject.org/coprs/maxamillion/origin-next/repo/epel-7/maxamillion-origin-next-epel-7.repo: dial tcp 209.132.184.54:443: no route to host
```
https://github.com/openshift/origin/issues/6040

File: test/cmd/images.sh [hunk 1]
```diff
@@ -72,7 +72,7 @@ os::cmd::expect_success_and_text 'oc describe istag/mysql:latest' 'Image Created
 os::cmd::expect_success_and_text 'oc describe istag/mysql:latest' 'Image Name:'
 name=$(oc get istag/mysql:latest --template='{{ .image.metadata.name }}')
 imagename="isimage/mysql@${name:0:15}"
-os::cmd::expect_success 'oc describe "${imagename}"'
+os::cmd::expect_success "oc describe ${imagename}"
```
Comment (liggitt) [inline test/cmd/images.sh]: do variables no longer get expanded in the expect functions, or was this just for consistency?

Comment (stevekuznetsov) [inline test/cmd/images.sh]: They do, this was just for consistency and so that the "name" of the year could pick up the value of the var.

Comment (stevekuznetsov) [inline test/cmd/images.sh]: In the README I suggest always double-quoting in the test file so that the output to the command line ("Running ..." or "SUCCESS:", etc) has the variables and you can debug better.

Comment (deads2k) [inline test/cmd/images.sh]: > In the README I suggest always double-quoting in the test file so that the output to the command line ("Running ..." or "SUCCESS:", etc) has the variables and you can debug better.

Add the readme to this commit?

Comment (stevekuznetsov) [inline test/cmd/images.sh]: @deads2k I think these two should be separate, as this PR is finished and ready as it is, the README has a long way to go yet. This PR is immediately useful and the `local` changes are potential bugs - something I'd rather go in now and not wait for the README.

**Comment by stevekuznetsov** [comment]
re[test]

File: hack/cmd_util.sh [hunk 2]
```diff
@@ -147,41 +147,50 @@ os_cmd_internal_tmperr="${os_cmd_internal_tmpdir}/tmp_stderr.log"
...
 	if [[ -n "${grep_args}" ]]; then
 		test_result=$( os::cmd::internal::run_collecting_output 'os::cmd::internal::get_results | grep -Eq "${grep_args}"'; echo $? )
 		
 	fi
-	test_succeeded=$( ${test_eval_func} "${test_result}"; echo $? )
+	local test_succeeded=$( ${test_eval_func} "${test_result}"; echo $? )
 
-	os::text::clear_last_line
+	local end_time=$(os::cmd::internal::seconds_since_epoch)
+	local time_elapsed=$(echo "scale=9; ${end_time} - ${start_time}" | bc | xargs printf '%11.9f') # we need leading zeroes for parsing later
```
Comment (deads2k) [inline hack/cmd_util.sh]: Add a comment saying what the intended output format is.  Since you said something about parsing, I'm guessing millis?

Comment (deads2k) [inline hack/cmd_util.sh]: > Add a comment saying what the intended output format is. Since you said something about parsing, I'm guessing millis?

Wait, floating point? kids these days.

File: hack/cmd_util.sh [hunk 3]
```diff
@@ -147,41 +147,50 @@ os_cmd_internal_tmperr="${os_cmd_internal_tmpdir}/tmp_stderr.log"
...
-	if (( cmd_succeeded && test_succeeded )); then
+	# some commands are multi-line, so we may need to clear more than just the previous line
+	local cmd_length=$(echo "${cmd}" | wc -l)
+	for (( i=0; i<${cmd_length}; i++ )); do
+		os::text::clear_last_line
+	done
 
-		os::text::print_green_bold "SUCCESS: ${cmd}"
+	if (( cmd_succeeded && test_succeeded )); then
+		os::text::print_green_bold "SUCCESS after ${time_elapsed}s: ${name}"
```
Comment (deads2k) [inline hack/cmd_util.sh]: Alright, we know that I hate colors, but bold green is ugly.  How about regular green and you can keep your bold red.

Comment (liggitt) [inline hack/cmd_util.sh]: Agree on regular green

Comment (stevekuznetsov) [inline hack/cmd_util.sh]: There are few things I will fight to the death about and unfortunately for you the bold color green is one of them.

**Comment by stevekuznetsov** [comment]
@deads2k addressed comments

File: hack/cmd_util.sh [hunk 4]
```diff
@@ -283,38 +361,50 @@ function os::cmd::internal::assemble_causes() {
...
+		local cmd_succeeded=$( ${cmd_eval_func} "${cmd_result}"; echo $? )
 		if (( cmd_succeeded )); then
 			break
 		fi
 		sleep "${interval}"
 	done
 
-	os::text::clear_last_line
+	local end_time=$(os::cmd::internal::seconds_since_epoch)
+	local time_elapsed=$(echo "scale=9; ${end_time} - ${start_time}" | bc | xargs printf '%11.9f') # in decimal seconds, we need leading zeroes for parsing later
```
Comment (deads2k) [inline hack/cmd_util.sh]: millis here too.

File: pkg/cmd/experimental/syncgroups/ad/augmented_ldapinterface.go [hunk 5]
```diff
@@ -60,6 +62,7 @@ func (e *AugmentedADLDAPInterface) GroupEntryFor(ldapGroupUID string) (*ldap.Ent
 		return nil, err
 	}
 
+	fmt.Printf("Augmented AD: query for group entry %q\n", ldapGroupUID)
```
Comment (deads2k) [inline pkg/cmd/experimental/syncgroups/ad/augmented_ldapinterface.go]: You can't output this here.  Anything going to system out will mess up piping ability.

File: pkg/cmd/experimental/syncgroups/rfc2307/ldapinterface.go [hunk 6]
```diff
@@ -93,10 +93,13 @@ func (e *LDAPInterface) ExtractMembers(ldapGroupUID string) ([]*ldap.Entry, erro
 // of the LDAPInterface first, then sending an LDAP query if the cache did not contain the entry.
 // This also satisfies the LDAPGroupGetter interface
 func (e *LDAPInterface) GroupEntryFor(ldapGroupUID string) (*ldap.Entry, error) {
+	fmt.Printf("RFC2307: Querying cache %v for group %q", e.cachedGroups, ldapGroupUID)
 	group, exists := e.cachedGroups[ldapGroupUID]
 	if exists {
+		fmt.Println("... group found")
```
Comment (deads2k) [inline pkg/cmd/experimental/syncgroups/rfc2307/ldapinterface.go]: Why are these changes here.  Just revert them for this pull.

File: test/cmd/images.sh (hunk 1 above)
Comment (deads2k) [inline test/cmd/images.sh]: > In the README I suggest always double-quoting in the test file so that the output to the command line ("Running ..." or "SUCCESS:", etc) has the variables and you can debug better.

We can start there, but I'm actually not sure we'll always want that.  We want stable names to track historical flakes.  If the names change because the command has resources that have generated names, then the test name would continually change.

Comment (stevekuznetsov) [inline test/cmd/images.sh]: *Ensuring* that your variables don't get expanded before they get passed in might be very hard - if you are trying to pass in a string that contains single-quotes, it's going to look ugly and not be very clear if we also have to use single-quotes around the entire thing to make sure nothing gets expanded.

The added benefit of seeing the actual command being run is pretty useful, too.

**Comment by stevekuznetsov** [comment]
Flake: https://github.com/openshift/origin/issues/6077

**Comment by stevekuznetsov** [comment]
re[test]

File: hack/cmd_util.sh [hunk 7]
```diff
@@ -147,41 +147,50 @@ os_cmd_internal_tmperr="${os_cmd_internal_tmpdir}/tmp_stderr.log"
...
+	local test_eval_func=${4:-os::cmd::internal::success_func}
 
 	os::cmd::internal::init_tempdir
 
-	echo "Running  ${cmd}..."
-	
-	cmd_result=$( os::cmd::internal::run_collecting_output "${cmd}"; echo $? )
-	cmd_succeeded=$( ${cmd_eval_func} "${cmd_result}"; echo $? )
+	local name=$(os::cmd::internal::describe_call "${cmd}" "${cmd_eval_func}" "${grep_args}" "${test_eval_func}")
+	echo "${name}..."
```
Comment (deads2k) [inline hack/cmd_util.sh]: Why did you remove "Running "?

**Comment by deads2k** [comment]
One last comment to fix up output (probably accidentally removed) and this lgtm.

Also, just connecting to a local API server and getting an answer back takes 170 milliseconds or so.  That's very disappointing.

File: hack/cmd_util.sh (hunk 7 above)
Comment (stevekuznetsov) [inline hack/cmd_util.sh]: The output as it reads now is more expressive and `Running` doesn't add anything:
```
test/cmd/export.sh:14: executing 'oc new-app -f examples/sample-app/application-template-stibuild.json --name=sample' expecting success...
```

**Comment by stevekuznetsov** [comment]
Probing what times the API server was taking was the impetus for @smarterclayton to add the 4k of REST calls to the tail of test-cmd, and, yes, they all are that bad or worse.

File: hack/cmd_util.sh (hunk 7 above)
Comment (deads2k) [inline hack/cmd_util.sh]: > The output as it reads now is more expressive and Running doesn't add anything:

Its more expressive to me.

Comment (liggitt) [inline hack/cmd_util.sh]: `os::cmd::internal::describe_call` builds a descriptive string that includes "executing ..."

Comment (stevekuznetsov) [inline hack/cmd_util.sh]: It didn't read well to me, but, sure, adding Running is fine to me.

**Comment by stevekuznetsov** [comment]
@deads2k reinstated `Running...`

**Comment by deads2k** [comment]
looks as good as bash ever looks.

[merge]

**Comment by smarterclayton** [comment]
Most of that cost is SSL establishment.  SSL is really, really slow to initiate.  CLIs will always be slow because of that.

**Comment by liggitt** [comment]
And version negotiation... don't forget we have two extra requests for every invocation
```
$ time oc get pods  --loglevel=6
I1204 15:57:54.833911    5575 loader.go:241] Config loaded from file /home/jliggitt/go/src/github.com/openshift/origin/openshift.local.config/master/admin.kubeconfig
I1204 15:57:54.834665    5575 loader.go:241] Config loaded from file /home/jliggitt/go/src/github.com/openshift/origin/openshift.local.config/master/admin.kubeconfig
I1204 15:57:54.980042    5575 debugging.go:121] GET https://192.168.1.114:8443/oapi 200 OK in 22 milliseconds
I1204 15:57:54.981119    5575 loader.go:241] Config loaded from file /home/jliggitt/go/src/github.com/openshift/origin/openshift.local.config/master/admin.kubeconfig
I1204 15:57:54.981876    5575 loader.go:241] Config loaded from file /home/jliggitt/go/src/github.com/openshift/origin/openshift.local.config/master/admin.kubeconfig
I1204 15:57:54.982789    5575 loader.go:241] Config loaded from file /home/jliggitt/go/src/github.com/openshift/origin/openshift.local.config/master/admin.kubeconfig
I1204 15:57:54.983536    5575 loader.go:241] Config loaded from file /home/jliggitt/go/src/github.com/openshift/origin/openshift.local.config/master/admin.kubeconfig
I1204 15:57:54.984382    5575 debugging.go:121] GET https://192.168.1.114:8443/api 200 OK in 0 milliseconds
I1204 15:57:54.985634    5575 debugging.go:121] GET https://192.168.1.114:8443/api/v1/namespaces/default/pods 200 OK in 0 milliseconds
```

the time spent serving an actual API call seems to be about 0-2ms after the initial setup

**Comment by smarterclayton** [comment]
We should start showing our response time in microseconds.

On Dec 4, 2015, at 3:58 PM, Jordan Liggitt <notifications@github.com> wrote:

And version negotiation... don't forget we have two extra requests for
every invocation

$ time oc get pods  --loglevel=6
I1204 15:57:54.833911    5575 loader.go:241] Config loaded from file
/home/jliggitt/go/src/github.com/openshift/origin/openshift.local.config/master/admin.kubeconfig
I1204 15:57:54.834665    5575 loader.go:241] Config loaded from file
/home/jliggitt/go/src/github.com/openshift/origin/openshift.local.config/master/admin.kubeconfig
I1204 15:57:54.980042    5575 debugging.go:121] GET
https://192.168.1.114:8443/oapi 200 OK in 22 milliseconds
I1204 15:57:54.981119    5575 loader.go:241] Config loaded from file
/home/jliggitt/go/src/github.com/openshift/origin/openshift.local.config/master/admin.kubeconfig
I1204 15:57:54.981876    5575 loader.go:241] Config loaded from file
/home/jliggitt/go/src/github.com/openshift/origin/openshift.local.config/master/admin.kubeconfig
I1204 15:57:54.982789    5575 loader.go:241] Config loaded from file
/home/jliggitt/go/src/github.com/openshift/origin/openshift.local.config/master/admin.kubeconfig
I1204 15:57:54.983536    5575 loader.go:241] Config loaded from file
/home/jliggitt/go/src/github.com/openshift/origin/openshift.local.config/master/admin.kubeconfig
I1204 15:57:54.984382    5575 debugging.go:121] GET
https://192.168.1.114:8443/api 200 OK in 0 milliseconds
I1204 15:57:54.985634    5575 debugging.go:121] GET
https://192.168.1.114:8443/api/v1/namespaces/default/pods 200 OK in 0
milliseconds

the time spent serving an actual API call seems to be about 0-2ms after the
initial setup

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/6133#issuecomment-162080328>.
